pkg/icingaredis/telemetry/heartbeat.go

```diff
@@ -146,5 +109,57 @@ func StartHeartbeat(
 	})
 }
 
-// Assert interface compliance.
-var _ json.Marshaler = jsonMetricValue{}
+type GoMetrics struct {
+	names   []string
+	units   []string
+	samples []metrics.Sample
+}
+
+func NewGoMetrics() *GoMetrics {
+	m := &GoMetrics{}
+
+	forbiddenRe := regexp.MustCompile(`\W`)
+
+	for _, d := range metrics.All() {
+		switch d.Kind {
+		case metrics.KindUint64, metrics.KindFloat64:
+			name := "go_" + strings.TrimPrefix(forbiddenRe.ReplaceAllString(d.Name, "_"), "_")
+
+			unit := ""
+			if strings.HasSuffix(d.Name, ":bytes") {
+				unit = "B"
+			} else if strings.HasSuffix(d.Name, ":seconds") {
+				unit = "s"
+			} else if d.Cumulative {
+				unit = "c"
+			}
+
+			m.names = append(m.names, name)
+			m.units = append(m.units, unit)
+			m.samples = append(m.samples, metrics.Sample{Name: d.Name})
+		}
+	}
+
+	return m
+}
+
+func (g *GoMetrics) PerformanceData() string {
+	metrics.Read(g.samples)
+
+	buf := bytes.NewBuffer(nil)
+
+	for i, value := range g.samples {
+		if i > 0 {
+			_, _ = fmt.Fprint(buf, " ")
+		}
+
+		switch value.Value.Kind() {
+		case metrics.KindUint64:
+			_, _ = fmt.Fprintf(buf, "%s=%d%s", g.names[i], value.Value.Uint64(), g.units[i])
+		case metrics.KindFloat64:
+			_, _ = fmt.Fprintf(buf, "%s=%f%s", g.names[i], value.Value.Float64(), g.units[i])
```

Comment:
_Should_ be deduplicated a bit.

Reply:
Do you see much potential for deduplication here that wouldn't break static checks for the format strings?

Reply:
Oh. 🐘